sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformHierarchy.java

```diff
@@ -32,32 +32,42 @@
 import java.util.Map.Entry;
 import java.util.Set;
 import javax.annotation.Nullable;
+import org.apache.beam.sdk.Pipeline;
 import org.apache.beam.sdk.Pipeline.PipelineVisitor;
 import org.apache.beam.sdk.Pipeline.PipelineVisitor.CompositeBehavior;
+import org.apache.beam.sdk.runners.PTransformOverrideFactory.ReplacementOutput;
 import org.apache.beam.sdk.transforms.AppliedPTransform;
 import org.apache.beam.sdk.transforms.PTransform;
 import org.apache.beam.sdk.values.PInput;
 import org.apache.beam.sdk.values.POutput;
 import org.apache.beam.sdk.values.PValue;
 import org.apache.beam.sdk.values.TaggedPValue;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * Captures information about a collection of transformations and their
  * associated {@link PValue}s.
  */
 public class TransformHierarchy {
+  private static final Logger LOG = LoggerFactory.getLogger(TransformHierarchy.class);
+
+  private final Pipeline pipeline;
   private final Node root;
+  private final Map<Node, PInput> unexpandedInputs;
```

Comment:
These are used only in `finishSpecifying`, yes - this probably should be split out, though